Implement XR_FB_color_space extension

devloglogan opened the pull request:
This PR adds support for the XR_FB_color_space extension.

Color space can be set in project settings or via the singleton at runtime, by default I have the color space set to Rec. 709. The reasoning (from my limited understanding of color space) is because Godot works in a sRGB / Rec. 709 color space.

~~I was torn on if this extension should be enabled by default. Setting the color space to Rec. 709 does slightly alter the color output, so those with existing projects may not want it enabled by default? I have it disabled for now, but could easily be swayed if others want it enabled.~~
2025-06-11 Edit: I've made it so that color space is enabled by default. ~~By default it looks like Quest 3 actually uses the Rec. 709 color space, so I don't think this will be problematic at all.~~
2025-06-12 Edit: Just kidding, it definitely uses a different color space than Rec. 709 enabled by default, so I've set the default color space to `Runtime Default`. This way nothing should change for any existing projects.

Also fixed a typo in `openxr_fb_render_model_extension_wrapper.cpp`.

dsnopek commented on plugin/src/main/cpp/include/extensions/openxr_fb_color_space_extension_wrapper.h:
```diff
@@ -0,0 +1,121 @@
+	};
+
+	Vector<XrColorSpaceFB> supported_color_spaces;
+	ColorSpace current_color_space;
```
```suggestion
	ColorSpace current_color_space = COLOR_SPACE_UNMANAGED;
```

Otherwise, this will contain garbage data until the OpenXR session starts.

devloglogan replied:
Updated this to `COLOR_SPACE_RUNTIME_DEFAULT`

dsnopek commented on plugin/src/main/cpp/extensions/openxr_fb_color_space_extension_wrapper.cpp:
```diff
@@ -0,0 +1,274 @@
+		return;
+	}
+
+	current_color_space = from_openxr_color_space(system_color_space_properties.colorSpace);
```
Hm, personally, I would have expected `current_color_space` to be unmanaged until explicitly set by the user, rather using the native color space. How do we want this to work?

Also, I think we should probably clear `current_color_space` when the OpenXR session ends, because I don't think the value is retained between sessions

devloglogan replied:
I've added `COLOR_SPACE_RUNTIME_DEFAULT` as a value.

Despite what the OpenXR spec says, the runtime default (at least for my Quest 3) is Rec 709, not Rift CV1. I am thinking that the runtime default and the device native color space are one in the same (this appears to be the case for my Quest 3), but maybe I'm incorrect here?

dsnopek replied:
> So it looks like if the default is not specified, then `system_color_space_properties.colorSpace` is the color space used when displaying the app at runtime.

If this was true, then we wouldn't see the subtle color difference we see when setting the color space to Rec 709 on Quest 3, because isn't its native color space Rec 709?

dsnopek commented on plugin/src/main/cpp/include/extensions/openxr_fb_color_space_extension_wrapper.h:
```diff
@@ -0,0 +1,121 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef OPENXR_FB_COLOR_SPACE_EXTENSION_WRAPPER_H
+#define OPENXR_FB_COLOR_SPACE_EXTENSION_WRAPPER_H
```
This is fine, but now that Godot and godot-cpp have switched to `#pragma once`, I've been trying to use that rather than traditional header guards in all new headers

dsnopek reviewed (commented):
Thanks!

Should we enable this extension in one (or maybe all?) of our sample projects? It is something that most folks should probably enable.

I did a little testing, and experienced a weird issue in the main demo project, where if I enabled the color space extension, then render models would stop working. I think it might be related to the `void **` vs `void *` thing that I pointed out.

Otherwise, it seems to be working! I made minor changes to the main demo project so I could change color spaces with a button press, and was able to notice minor differences.

devloglogan commented on doc_classes/OpenXRFbColorSpaceExtensionWrapper.xml:
```diff
@@ -0,0 +1,70 @@
+	</brief_description>
+	<description>
+		Wraps the [code]XR_FB_color_space[/code] extension.
+	</description>
```
I've expanded upon it a bit!

devloglogan commented on plugin/src/main/cpp/extensions/openxr_fb_color_space_extension_wrapper.cpp:
```diff
@@ -0,0 +1,274 @@
+		return;
+	}
+
+	supported_color_spaces_set = true;
```
I think that makes sense, I've added it.

m4gr3d commented on plugin/src/main/cpp/extensions/openxr_fb_color_space_extension_wrapper.cpp:
```diff
@@ -0,0 +1,291 @@
+		} break;
+		case COLOR_SPACE_RUNTIME_DEFAULT:
+		default: {
+			ERR_FAIL_V_MSG(XR_COLOR_SPACE_UNMANAGED_FB, vformat("Unknown color space: %s", p_color_space));
```
We shouldn't default to `XR_COLOR_SPACE_UNMANAGED_FB` since this applies different settings than the runtime default.
Can we just return `XR_COLOR_SPACE_MAX_ENUM_FB` instead?

m4gr3d commented on plugin/src/main/cpp/register_types.cpp:
```diff
@@ -417,6 +425,22 @@ void add_plugin_project_settings() {
+	{
+		String starting_color_space = "xr/openxr/extensions/meta/color_space/starting_color_space";
+		if (!project_settings->has_setting(starting_color_space)) {
+			project_settings->set_setting(starting_color_space, 2);
```
The default value should be `Runtime Default`

devloglogan replied:
I've swapped the default to `Runtime Default`!

m4gr3d commented on plugin/src/main/cpp/register_types.cpp:
```diff
@@ -417,6 +425,22 @@ void add_plugin_project_settings() {
+		property_info["name"] = starting_color_space;
+		property_info["type"] = Variant::Type::INT;
+		property_info["hint"] = PROPERTY_HINT_ENUM;
+		property_info["hint_string"] = "Unmanaged:0,REC2020:1,REC709:2,Rift CV1:3,Rift S:4,Quest:5,P3:6,Adobe RGB:7,Runtime Default:8";
```
`Runtime Default` should be zeroth index value.

dsnopek commented on plugin/src/main/cpp/extensions/openxr_fb_color_space_extension_wrapper.cpp:
```diff
@@ -0,0 +1,291 @@
+		return;
+	}
+
+	current_color_space = from_openxr_color_space(system_color_space_properties.colorSpace);
```
Wait, I'm still not sure this is right. Isn't `system_color_space_properties.colorSpace` the device's native colorspace and not the current color space?

The Quest 3 should be returning Rec 709 here, because that's its native colorspace. But we still see a subtle color difference when we set the colorspace to Rec 709. Which leads me to think that it doesn't necessary start on the native color space, but could be some other value.

Now that `set_color_space()` has this code:

```
	if (p_color_space == current_color_space) {
		return;
	}
```

... I think setting the color space to Rec 709 would fail on the Quest 3, because `current_color_space` is Rec 709.

Perhaps we shouldn't even have this line? Then it won't change the `current_color_space` on session start, and it would still be "runtime default" until we explicitly set it

devloglogan replied:
Huh I thought I'd maybe made an incorrect observation before, and confirmed that the native color space was the runtime default (at least on Quest 3), but you're definitely correct. I've removed the session created function override entirely, so I don't believe this is a problem anymore!

This would have been a lot simpler if OpenXR gave us a way to check what color space is active; but, alas. 🙂

dsnopek reviewed (commented):
The code for this is looking great to me now! I retested, all seems to be working as expected.

I've gone through the docs, and attempted to clarify them for my understanding of how this extension works (which I hope is correct :-)).

I've been debating whether or not it makes sense to specifically say that at the moment Godot can only render in Rec 709, so the only rational options are either (1) Rec 709 for correctness, or (2) "Runtime Default" for existing projects that don't like how using Rec 709 changes their colors.

But I'm not sure the best way to say that, and it also may not be true forever (Godot may one day be able to render Rec 2020 as part of HDR support?), so I didn't include it in my suggestions. Might be better for a tutorial? Or, if you have a good idea for how to integrate, feel free to

I also think we should update all our samples to use this extension with the color space set to Rec 709, but we don't necessarily have to do that in this PR

devloglogan commented:
@dsnopek thanks for the docs suggestions, I've applied them. 🙂

m4gr3d reviewed (approved):
Looks good! 

Quick reminder about adding the warning for 4.x

devloglogan commented:
Added a warning on startup if the project setting isn't set to Rec. 709, does this work for you @m4gr3d?

I also updated some of the other error printing.

m4gr3d commented on plugin/src/main/cpp/extensions/openxr_fb_color_space_extension_wrapper.cpp:
```diff
@@ -0,0 +1,275 @@
+
+	ColorSpace starting_color_space = ColorSpace((int)project_settings->get_setting_with_override("xr/openxr/extensions/meta/color_space/starting_color_space"));
+	if (starting_color_space != COLOR_SPACE_REC709) {
+		WARN_PRINT("Recommended color space project setting is REC709");
```
Looks good!

Alongside this warning, let's also include a warning in the export window by updating https://github.com/GodotVR/godot_openxr_vendors/blob/001d03c087c3761b96debceb037e7be3e366eaa7/plugin/src/main/cpp/export/export_plugin.cpp#L184

dsnopek replied:
It could be a little confusing getting a warning about a project setting, when in the export settings area - folks could start looking in the wrong place

devloglogan replied:
I agree, feels strange to have a warning about project settings in export settings.

m4gr3d replied:
I disagree; with the switch to gating the extensions via the project settings, and moving related export options to project settings, the two are intrinsically linked. 
Taking a look at `meta_export_plugin.cpp#_get_export_option_warning(...)` most of them check a project setting in order to dispatch a warning. I see this as following a similar approach.

The export window is also the most effective place to surface the warning (up until we update the default in `v5`), because a warning in the logs can be easily missed, but here we prompt you, right before you are about to export your game, to ensure it looks correct by selecting the adequate color space.

If there is concern about folks looking in the wrong place, let's just be explicit about how they can address the warning. We do the same thing with the export option, where we specify the name of the export option that needs to be udpated.

devloglogan replied:
@m4gr3d @dsnopek ultimately I'm fine with either way, and  it's true that without adding another warning somewhere, it's entirely possible developers would miss the warning entirely.

I've added a warning to the export window as well. Let me know if you find this acceptable!
